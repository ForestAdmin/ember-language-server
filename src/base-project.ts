import { getPodModulePrefix } from './utils/layout-helpers';
import { ClassicPathMatcher, PodMatcher } from './utils/path-matcher';

export class BaseProject {
  private classicMatcher!: ClassicPathMatcher;
  private podMatcher!: PodMatcher;
  podModulePrefix = '';
  constructor(public readonly root: string) {
    const maybePrefix = getPodModulePrefix(root);

    if (maybePrefix) {
      this.podModulePrefix = 'app/' + maybePrefix;
    } else {
      this.podModulePrefix = 'app';
    }

    this.classicMatcher = new ClassicPathMatcher(this.root);
    this.podMatcher = new PodMatcher(this.root, this.podModulePrefix);
  }
  matchPathToType(filePath: string) {
    const item = this.classicMatcher.metaFromPath(filePath) || this.podMatcher.metaFromPath(filePath);

    if (item && filePath.toLowerCase().includes('/app/features/')) {
      item.name = 'feature' + filePath.split('/app/features')[1].replace('/components/', '/');
      item.name = item.name.split('/').slice(0, -1).join('/');
    }

    if (item && filePath.toLowerCase().includes('/app/shared/')) {
      item.name = 'shared/' + item.name;
    }

    return item;
  }
}
