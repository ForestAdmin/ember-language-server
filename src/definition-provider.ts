import * as path from 'path';
import * as fs from 'fs';

import {
  RequestHandler,
  TextDocumentPositionParams,
  Definition,
  Location,
  Range
} from 'vscode-languageserver';

import { parse } from 'babylon';

import { toPosition } from './estree-utils';
import Server from './server';
import ASTPath from './glimmer-utils';
import { getExtension } from './utils/file-extension';
import { getProjectAddonsRoots, getPodModulePrefix } from './completion-provider/template-completion-provider';
import URI from 'vscode-uri';
const _ = require('lodash');
const memoize = require('memoizee');

const mProjectAddonsRoots = memoize(getProjectAddonsRoots, {
  length: 1,
  maxAge: 600000
});
const mAddonPathsForComponentTemplates = memoize(
  getAddonPathsForComponentTemplates,
  { length: 2, maxAge: 600000 }
);

const { preprocess } = require('@glimmer/syntax');

function getAddonPathsForComponentTemplates(
  root: string,
  maybeComponentName: string
) {
  const roots = mProjectAddonsRoots(root);
  let existingPaths: string[] = [];
  let hasValidPath = false;
  roots.forEach((rootPath: string) => {
    if (hasValidPath) {
      return;
    }
    const addonPaths: string[][] = [];

    getAbstractComponentScriptsParts(rootPath, 'addon', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });
    getAbstractComponentScriptsParts(rootPath, 'app', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });
    getAbstractComponentTemplatesParts(rootPath, 'app', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });
    getAbstractComponentTemplatesParts(rootPath, 'addon', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });
    getAbstractHelpersParts(rootPath, 'app', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });
    getAbstractHelpersParts(rootPath, 'addon', maybeComponentName).forEach((parts: any) => {
      addonPaths.push(parts);
    });

    const validPaths = addonPaths
      .map((pathArr: string[]) => {
        return path.join.apply(path, pathArr.filter((part: any) => !!part));
      })
      .filter(fs.existsSync);
    if (validPaths.length) {
      hasValidPath = true;
      existingPaths = validPaths;
    }
  });

  const addonFolderFiles =
    existingPaths.filter(hasAddonFolderInPath);
  if (addonFolderFiles.length) {
    return addonFolderFiles;
  }
  return existingPaths;
}

function getPathsForComponentTemplates(
  root: string,
  maybeComponentName: string
): string[] {
  const podModulePrefix = getPodModulePrefix(root);
  let podComponentsScriptsParts: string[][] = [];
  if (podModulePrefix) {
    podComponentsScriptsParts = getAbstractComponentScriptsParts(root, 'app' + path.sep + podModulePrefix, maybeComponentName);
  }
  const paths = [...podComponentsScriptsParts, ...getAbstractComponentTemplatesParts(root, 'app', maybeComponentName)]
  .map((pathParts: any) => {
    return path.join.apply(path, pathParts.filter((part: any) => !!part));
  });
  return paths;
}

function getAbstractComponentScriptsParts(root: string, prefix: string, maybeComponentName: string) {
  return [
    [root, prefix, 'components', maybeComponentName + '.js'],
    [root, prefix, 'components', maybeComponentName + '.ts'],
    [root, prefix, 'components', maybeComponentName, 'component.js'],
    [root, prefix, 'components', maybeComponentName, 'component.ts']
  ];
}

function getAbstractComponentTemplatesParts(root: string, prefix: string, maybeComponentName: string) {
  return [
    [root, prefix, 'components', maybeComponentName, 'template.hbs'],
    [root, prefix, 'templates', 'components', maybeComponentName + '.hbs']
  ];
}

function getAbstractHelpersParts(root: string, prefix: string, maybeComponentName: string) {
  return [
    [root, prefix, 'helpers', `${maybeComponentName}.js`],
    [root, prefix, 'helpers', `${maybeComponentName}.ts`]
  ];
}

function getPathsForComponentScripts(
  root: string,
  maybeComponentName: string
): string[] {
  const podModulePrefix = getPodModulePrefix(root);
  let podComponentsScriptsParts: string[][] = [];
  if (podModulePrefix) {
    podComponentsScriptsParts = getAbstractComponentScriptsParts(root, 'app' + path.sep + podModulePrefix, maybeComponentName);
  }
  const paths = [...podComponentsScriptsParts, ...getAbstractComponentScriptsParts(root, 'app', maybeComponentName)].map((pathParts: any) => {
    return path.join.apply(path, pathParts.filter((part: any) => !!part));
  });
  return paths;
}

function getComponentNameFromURI(root: string, uri: string) {
  let fileName = uri.replace('file://', '').replace(root, '');
  let maybeComponentName = fileName
    .split(path.sep)
    .join('/')
    .split('/components/')[1];
  if (maybeComponentName.endsWith('/template.hbs')) {
    maybeComponentName = maybeComponentName.replace('/template.hbs', '');
  } else if (isTemplatePath(maybeComponentName)) {
    maybeComponentName = maybeComponentName.replace('.hbs', '');
  }
  return maybeComponentName;
}

export default class DefinitionProvider {
  constructor(private server: Server) {}

  handle(params: TextDocumentPositionParams): Definition | null {
    let uri = params.textDocument.uri;

    const project = this.server.projectRoots.projectForUri(uri);
    if (!project) {
      return null;
    }

    let extension = getExtension(params.textDocument);

    if (extension === '.hbs') {
      let content = this.server.documents.get(uri).getText();
      let ast = preprocess(content);
      let focusPath = ASTPath.toPosition(ast, toPosition(params.position));
      if (!focusPath) {
        return null;
      }

      if (this.isAngleComponent(focusPath)) {
        const maybeComponentName = _.kebabCase(focusPath.node.tag);

        let paths = [
          ...getPathsForComponentScripts(project.root, maybeComponentName),
          ...getPathsForComponentTemplates(project.root, maybeComponentName)
        ].filter(fs.existsSync);

        if (!paths.length) {
          paths = mAddonPathsForComponentTemplates(
            project.root,
            maybeComponentName
          );
        }

        return pathsToLocations.apply(
          null,
          paths.length > 1
            ? paths.filter((postfix: string) => isTemplatePath(postfix))
            : paths
        );
      } else if (this.isComponentWithBlock(focusPath)) {
        let maybeComponentName = focusPath.node.path.original;
        let paths: string[] = getPathsForComponentTemplates(
          project.root,
          maybeComponentName
        ).filter(fs.existsSync);
        if (!paths.length) {
          paths = mAddonPathsForComponentTemplates(
            project.root,
            maybeComponentName
          ).filter((name: string) => {
            return isTemplatePath(name);
          });
        }
        // mAddonPathsForComponentTemplates
        return pathsToLocationsWithPosition(paths, '{{yield');
      } else if (
        this.isActionName(focusPath) ||
        this.isLocalProperty(focusPath) ||
        this.isHashPairWithLocalValue(focusPath)
      ) {
        let maybeComponentName = getComponentNameFromURI(project.root, uri);
        let paths: string[] = getPathsForComponentScripts(
          project.root,
          maybeComponentName
        ).filter(fs.existsSync);
        if (!paths.length) {
          paths = mAddonPathsForComponentTemplates(
            project.root,
            maybeComponentName
          ).filter((name: string) => {
            return !isTemplatePath(name);
          });
        }
        const text =
          focusPath.node.type !== 'HashPair'
            ? focusPath.node.original
            : focusPath.node.value.original;
        return pathsToLocationsWithPosition(
          paths,
          text.replace('this.', '').split('.')[0]
        );
      } else if (this.isComponentOrHelperName(focusPath)) {
        const maybeComponentName =
          focusPath.node.type === 'ElementNode'
            ? _.kebabCase(focusPath.node.tag)
            : focusPath.node.original;

        let helpers = getAbstractHelpersParts(project.root, 'app', maybeComponentName)
        .map((pathParts: any) => {
          return path.join.apply(path, pathParts.filter((part: any) => !!part));
        });

        let paths = [
          ...getPathsForComponentScripts(project.root, maybeComponentName),
          ...getPathsForComponentTemplates(project.root, maybeComponentName),
          ...helpers
        ].filter(fs.existsSync);

        if (!paths.length) {
          paths = mAddonPathsForComponentTemplates(
            project.root,
            maybeComponentName
          );
        }

        return pathsToLocations.apply(
          null,
          paths.length > 1
            ? paths.filter(isTemplatePath)
            : paths
        );
      } else {
        // let { line, column } =  toPosition(params.position);
        // let textLine = getLineFromText(content, line);
        // let leftLine = textLine.slice(0, column);
        // let rightLine = textLine.slice(column);
        // let leftEqual = leftLine.lastIndexOf('=');
        // if (leftEqual > -1) {
        //   let needle = leftLine.slice(leftEqual + 1).trim() + rightLine.slice(0, rightLine.indexOf(' ') || rightLine.length - 1).trim();
        //   if (needle.indexOf('this.') > -1) {
        //   }
        // }
      }
    } else if (extension === '.js') {
      let content = this.server.documents.get(uri).getText();
      let ast = parse(content, {
        sourceType: 'module'
      });
      let astPath = ASTPath.toPosition(ast, toPosition(params.position));
      if (!astPath) {
        return null;
      }

      if (isModelReference(astPath)) {
        let modelName = astPath.node.value;
        const modelPath = path.join(
          project.root,
          'app',
          'models',
          `${modelName}.js`
        );
        const tsModelPath = path.join(
          project.root,
          'app',
          'models',
          `${modelName}.ts`
        );
        return pathsToLocations(modelPath, tsModelPath);
      } else if (isTransformReference(astPath)) {
        let transformName = astPath.node.value;

        const transformPath = path.join(
          project.root,
          'app',
          'transforms',
          `${transformName}.js`
        );
        const tsTransformPath = path.join(
          project.root,
          'app',
          'transforms',
          `${transformName}.ts`
        );
        return pathsToLocations(transformPath, tsTransformPath);
      }
    }

    return null;
  }

  get handler(): RequestHandler<TextDocumentPositionParams, Definition, void> {
    return this.handle.bind(this);
  }

  isLocalProperty(path: ASTPath) {
    let node = path.node;
    if (node.type === 'PathExpression') {
      return node.this;
    }
    return false;
  }

  isHashPairWithLocalValue(path: ASTPath) {
    let node = path.node;
    return (
      node.type === 'HashPair' &&
      node.value.type === 'PathExpression' &&
      node.value.this
    );
  }

  isActionName(path: ASTPath) {
    let node = path.node;
    if (path.parent.type !== 'PathExpression') {
      return false;
    }
    if (
      !path.parent ||
      path.parent.path.original !== 'action' ||
      !path.parent.params[0] === node
    ) {
      return false;
    }
    if (node.type === 'StringLiteral') {
      return true;
    } else if (node.type === 'PathExpression' && node.this) {
      return true;
    }
    return false;
  }

  isComponentWithBlock(path: ASTPath) {
    let node = path.node;
    return (
      node.type === 'BlockStatement' &&
      node.path.type === 'PathExpression' &&
      node.path.this === false &&
      node.path.original.includes('-') &&
      node.path.original.charAt(0) !== '-' &&
      !node.path.original.includes('.')
    );
  }

  isAngleComponent(path: ASTPath) {
    let node = path.node;

    if (node.type === 'ElementNode') {
      if (node.tag.charAt(0) === node.tag.charAt(0).toUpperCase()) {
        return true;
      }
    }
  }

  isComponentOrHelperName(path: ASTPath) {
    let node = path.node;

    if (this.isAngleComponent(path)) {
      return true;
    }

    if (node.type === 'StringLiteral') {
      // if (node.original.includes('/')) {
      //   return true;
      // } else if (!node.original.includes('.') && node.original.includes('-')) {
      //   return true;
      // }
      if (
        path.parent &&
        path.parent.path.original === 'component' &&
        path.parent.params[0] === node
      ) {
        return true;
      }
    }

    if (node.type !== 'PathExpression') {
      return false;
    }

    let parent = path.parent;
    if (
      !parent ||
      parent.path !== node ||
      (parent.type !== 'MustacheStatement' &&
        parent.type !== 'BlockStatement' &&
        parent.type !== 'SubExpression')
    ) {
      return false;
    }

    return true;
  }
}

function isModelReference(astPath: ASTPath): boolean {
  let node = astPath.node;
  if (node.type !== 'StringLiteral') {
    return false;
  }
  let parent = astPath.parent;
  if (
    !parent ||
    parent.type !== 'CallExpression' ||
    parent.arguments[0] !== node
  ) {
    return false;
  }
  let identifier =
    parent.callee.type === 'Identifier'
      ? parent.callee
      : parent.callee.property;
  return identifier.name === 'belongsTo' || identifier.name === 'hasMany';
}

function isTransformReference(astPath: ASTPath): boolean {
  let node = astPath.node;
  if (node.type !== 'StringLiteral') {
    return false;
  }
  let parent = astPath.parent;
  if (
    !parent ||
    parent.type !== 'CallExpression' ||
    parent.arguments[0] !== node
  ) {
    return false;
  }
  let identifier =
    parent.callee.type === 'Identifier'
      ? parent.callee
      : parent.callee.property;
  return identifier.name === 'attr';
}

function pathsToLocations(...paths: string[]): Location[] {
  return paths.filter(fs.existsSync).map(modulePath => {
    return Location.create(
      URI.file(modulePath).toString(),
      Range.create(0, 0, 0, 0)
    );
  });
}

function isTemplatePath(filePath: string) {
  return filePath.endsWith('.hbs');
}

function hasAddonFolderInPath(name: string) {
  return name.includes(path.sep + 'addon' + path.sep);
}

// function getLineFromText(text: string, line: number) {
//   const arrayOfLines = text.match(/[^\r\n]+/g) || [];
//   return arrayOfLines[line - 1] || '';
// }

function getFirstTextPostion(filePath: string, content: string) {
  const text = fs.readFileSync(filePath, 'utf8');
  const arrayOfLines = text.match(/(.*?(?:\r\n?|\n|$))/gm) || [];
  let startLine = 0;
  let startCharacter = 0;
  arrayOfLines.forEach((line: string, index: number) => {
    if (startLine || startCharacter) {
      return;
    }
    let textPosition = line.indexOf(content);
    let bounds = line.split(content);
    if (textPosition > -1) {
      if (
        /\s/.test(bounds[0].charAt(bounds[0].length - 1)) ||
        bounds[0].trim().length === 0
      ) {
        if (/^[A-Za-z]+$/.test(bounds[1].charAt(0)) === false) {
          startLine = index;
          startCharacter = textPosition;
        }
      }
    }
  });
  return [startLine, startCharacter];
}

function pathsToLocationsWithPosition(paths: string[], findMe: string) {
  return paths.filter(fs.existsSync).map((fileName: string) => {
    const [startLine, startCharacter] = getFirstTextPostion(fileName, findMe);
    return Location.create(
      URI.file(fileName).toString(),
      Range.create(
        startLine,
        startCharacter,
        startLine,
        startCharacter + findMe.length
      )
    );
  });
}
