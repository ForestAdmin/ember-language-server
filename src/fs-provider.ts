import * as fs from 'fs';
import { DocumentUri } from 'vscode-languageserver-protocol';
import { URI } from 'vscode-uri';

let currentFSImplementation!: FSProvider;

export function setFSImplementation(fs: FSProvider) {
  currentFSImplementation = fs;
}

export function fsProvider(): FSProvider {
  if (!currentFSImplementation) {
    setFSImplementation(new FSProvider());
  }

  return currentFSImplementation;
}

export default class FSProvider {
  // expected VSCode api, replacement of existsSync
  async exists(uri: DocumentUri | fs.PathLike): Promise<boolean> {
    const entry = URI.isUri(uri) ? URI.parse(uri as DocumentUri).fsPath : uri;

    try {
      await fs.statSync(entry);

      return true;
    } catch (e) {
      return false;
    }
  }
  // expected VSCode api, replacement of readFileSync
  async readFile(uri: DocumentUri | fs.PathLike): Promise<string> {
    const entry = URI.isUri(uri) ? URI.parse(uri as DocumentUri).fsPath : uri;
    const item = fs.readFileSync(entry, null);

    return item.toString('utf8');
  }
  // logger api
  createWriteStream(filePath: fs.PathLike, flags: { flags: string }) {
    return fs.createWriteStream(filePath, flags);
  }
  // walk-sync api
  statSync(filePath: fs.PathLike) {
    return fs.statSync(filePath);
  }
  realpathSync(filePath: fs.PathLike, options?: { encoding?: BufferEncoding | null } | BufferEncoding | null) {
    return fs.realpathSync(filePath, options);
  }
  readdirSync(filePath: fs.PathLike, options?: BufferEncoding | { encoding: BufferEncoding | null; withFileTypes?: false | undefined } | null | undefined) {
    return fs.readdirSync(filePath, options);
  }
}
