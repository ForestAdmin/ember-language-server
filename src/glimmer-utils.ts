import { Position, SourceLocation } from 'estree';
import { containsPosition } from './estree-utils';

type ScopeValue = [string, ASTPath, number];

export function componentNameForPath(astPath: ASTPath) {
  if (isLocalPathExpression(astPath)) {
    const scope = getLocalScope(astPath);
    const pathName = getLocalPathName(astPath.node);
    if (pathName) {
      let declaration = scope.find(([name]) => name === pathName);
      if (!declaration) {
        return;
      }
      if (declaration[1].node.type === 'ElementNode') {
        return declaration[1].node.tag;
      }
    }
  }
}

function getLocalPathName(node: any) {
  if (!node || node.type !== 'PathExpression' || !node.parts.length) {
    return undefined;
  }
  const pathName: string = node.parts[0];
  if (pathName === 'this') {
    return undefined;
  }
  return pathName;
}

export function isLocalPathExpression(astPath: ASTPath) {
  const pathName = getLocalPathName(astPath.node);
  if (!pathName) {
    return false;
  }
  const scope = getLocalScope(astPath);
  const declarations = scope.filter(([name]) => name === pathName);
  if (declarations.length) {
    return true;
  } else {
    return false;
  }
}

export function getLocalScope(astPath: ASTPath) {
  const scopeValues: ScopeValue[] = [];
  let cursor: ASTPath | undefined = astPath.parentPath;
  while (cursor) {
    const node = cursor.node;
    if (node && (node.type === 'ElementNode' || node.type === 'Block')) {
      const params = node.blockParams;
      params.forEach((param: string) => {
        scopeValues.push([param, cursor as ASTPath, params.indexOf(param)]);
      });
    }
    cursor = cursor.parentPath;
  }
  return scopeValues;
}
export default class ASTPath {
  static toPosition(ast: any, position: Position): ASTPath | undefined {
    let path = _findFocusPath(ast, position);
    if (path) {
      return new ASTPath(path);
    }
  }

  private constructor(private readonly path: any[], private readonly index: number = path.length - 1) {}

  get node(): any {
    return this.path[this.index];
  }

  get parent(): any | undefined {
    return this.path[this.index - 1];
  }

  get parentPath(): ASTPath | undefined {
    if (this.index - 1 < 0) {
      return undefined;
    }
    return new ASTPath(this.path, this.index - 1);
  }
}

function _findFocusPath(node: any, position: Position, seen = new Set()): any {
  seen.add(node);

  let path: any[] = [];
  let range: SourceLocation = node.loc;
  if (range) {
    if (containsPosition(range, position)) {
      path.push(node);
    } else {
      return [];
    }
  }

  for (let key in node) {
    if (!Object.prototype.hasOwnProperty.call(node, key)) {
      continue;
    }

    let value = node[key];
    if (!value || typeof value !== 'object' || seen.has(value)) {
      continue;
    }

    let childPath = _findFocusPath(value, position, seen);
    if (childPath.length > 0) {
      path = path.concat(childPath);
      break;
    }
  }

  return path;
}
