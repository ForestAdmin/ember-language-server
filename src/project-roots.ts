'use strict';

import * as path from 'path';
import { uriToFilePath } from 'vscode-languageserver/lib/files';
import { logError, logInfo } from './utils/logger';
import * as walkSync from 'walk-sync';
import { isGlimmerNativeProject, isGlimmerXProject } from './utils/layout-helpers';
import { ProjectProviders, collectProjectProviders, initBuiltinProviders } from './utils/addon-api';
import Server from './server';

export interface Executors {
  [key: string]: (server: Server, command: string, args: any[]) => any;
}

export class Project {
  providers!: ProjectProviders;
  builtinProviders!: ProjectProviders;
  executors: Executors = {};
  constructor(public readonly root: string, server: Server) {
    this.providers = collectProjectProviders(root);
    this.builtinProviders = initBuiltinProviders();
    this.providers.initFunctions.forEach((initFn) => initFn(server, this));
  }
}

export default class ProjectRoots {
  constructor(private server: Server) {}
  workspaceRoot: string;

  projects = new Map<string, Project>();

  findProjectsInsideRoot(workspaceRoot: string) {
    const roots = walkSync(workspaceRoot, {
      directories: false,
      globs: ['**/ember-cli-build.js', '**/package.json'],
      ignore: ['**/.git/**', '**/bower_components/**', '**/dist/**', '**/node_modules/**', '**/tmp/**']
    });

    roots.forEach((rootPath: string) => {
      const filePath = path.join(workspaceRoot, rootPath);
      const fullPath = path.dirname(filePath);
      if (filePath.endsWith('package.json')) {
        try {
          if (isGlimmerNativeProject(fullPath) || isGlimmerXProject(fullPath)) {
            this.onProjectAdd(fullPath);
          }
        } catch (e) {
          logError(e);
        }
      } else {
        this.onProjectAdd(fullPath);
      }
    });
  }

  async initialize(workspaceRoot: string) {
    this.workspaceRoot = workspaceRoot;

    this.findProjectsInsideRoot(this.workspaceRoot);
  }

  onProjectAdd(path: string) {
    if (this.projects.has(path)) {
      return;
    }
    try {
      this.projects.set(path, new Project(path, this.server));
      logInfo(`Ember CLI project added at ${path}`);
    } catch (e) {
      logError(e);
    }
  }

  projectForUri(uri: string): Project | undefined {
    let path = uriToFilePath(uri);

    if (!path) return;

    let root = (Array.from(this.projects.keys()) || []).filter((root) => path!.indexOf(root) === 0).reduce((a, b) => (a.length > b.length ? a : b), '');

    return this.projects.get(root);
  }
}
