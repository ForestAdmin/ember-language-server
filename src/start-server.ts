import { Server } from '.';
import { IPCMessageReader, IPCMessageWriter, StreamMessageWriter, StreamMessageReader, createConnection, Connection } from 'vscode-languageserver/node';

// Create a connection for the server. The connection defaults to Node's IPC as a transport, but
// also supports stdio via command line flag
const connection: Connection = process.argv.includes('--stdio')
  ? createConnection(new StreamMessageReader(process.stdin), new StreamMessageWriter(process.stdout))
  : createConnection(new IPCMessageReader(process), new IPCMessageWriter(process));

const server = new Server(connection);

server.listen();
