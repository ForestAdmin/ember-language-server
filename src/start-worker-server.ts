import { Server } from '.';
import { Connection, createConnection, BrowserMessageReader, BrowserMessageWriter } from 'vscode-languageserver/browser';

// require.include('.foo.js');
// require.include("@babel/core");
// require.include("@babel/plugin-proposal-decorators");
// require.include("@babel/plugin-proposal-class-properties");
// require.include("@babel/plugin-proposal-object-rest-spread");
// require.include("@babel/plugin-proposal-optional-chaining");
// require.include("@babel/plugin-proposal-nullish-coalescing-operator");
// require.include("@babel/plugin-proposal-numeric-separator");
// require.include("@babel/plugin-proposal-logical-assignment-operators");
// require.include("@babel/plugin-proposal-private-methods");
// require.include("@babel/plugin-proposal-async-generator-functions");
// require.include("@babel/plugin-proposal-function-sent");
// require.include("@babel/plugin-proposal-do-expressions");
// Create a connection for the server. The connection defaults to Node's IPC as a transport, but
// also supports stdio via command line flag
const connection: Connection = createConnection(new BrowserMessageReader(self), new BrowserMessageWriter(self));

const server = new Server(connection, {
  type: 'worker',
  fs: 'async',
});

server.listen();
