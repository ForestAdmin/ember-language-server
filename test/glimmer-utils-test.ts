import { Position } from 'vscode-languageserver';
import { preprocess } from '@glimmer/syntax';

import ASTPath, { getLocalScope, componentNameForPath, sourceForNode } from '../src/glimmer-utils';
import { toPosition } from '../src/estree-utils';

describe('glimmer-utils', function() {
  describe('ASTPath', function() {
    it('works as expected', function() {
      const input = `
<Component as |items|>
{{#let items as |item bar|}}
{{foo}}
{{/let}}
</Component>
        `;
      const astPath = ASTPath.toPosition(preprocess(input), toPosition(Position.create(3, 5)));
      expect(astPath.node).toMatchSnapshot();
    });
  });
  describe('getLocalScope', function() {
    it('works as expected', function() {
      const input = `
<Component as |items|>
{{#let items as |item bar|}}
{{foo}}
{{/let}}
</Component>
        `;
      const astPath = ASTPath.toPosition(preprocess(input), toPosition(Position.create(3, 5)));
      expect(getLocalScope(astPath).map(([el, , ind]) => [el, ind])).toEqual([['item', 0], ['bar', 1], ['items', 0]]);
    });
  });
  describe('componentNameForPath', function() {
    it('works as expected', function() {
      const input = `
<Component as |items|>
{{#let items as |item bar|}}
{{items}}
{{/let}}
</Component>
        `;
      const astPath = ASTPath.toPosition(preprocess(input), toPosition(Position.create(3, 5)));
      expect(componentNameForPath(astPath)).toEqual('Component');
    });
  });
  describe('sourceForNode', function() {
    it('works as expected', function() {
      const input = `
      <Component as |items|>
      {{#let items as |item bar|}}
      {{items}}
      {{/let}}
      </Component>
              `;
      const astPath = ASTPath.toPosition(preprocess(input), toPosition(Position.create(2, 2)));
      expect(astPath.node.tag).toEqual('Component');
      expect(sourceForNode(astPath.node, input)).toEqual(input.trim());
    });
    it('works as expected for MustachePaths', function() {
      const input = ['<Component as |items|>', '{{#let items as |item bar|}}', '{{items}}', '{{/let}}', '</Component>'].join('\n');
      const astPath = ASTPath.toPosition(preprocess(input), toPosition(Position.create(2, 3)));
      expect(astPath.node.original).toEqual('items');
      expect(sourceForNode(astPath.node, input)).toEqual('items');
    });
  });
});
