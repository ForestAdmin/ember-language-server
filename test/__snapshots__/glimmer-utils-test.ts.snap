// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`glimmer-utils ASTPath works as expected 1`] = `
Object {
  "data": false,
  "loc": SourceLocation {
    "end": Object {
      "column": 5,
      "line": 4,
    },
    "source": undefined,
    "start": Object {
      "column": 2,
      "line": 4,
    },
  },
  "original": "foo",
  "parts": Array [
    "foo",
  ],
  "this": false,
  "type": "PathExpression",
}
`;

exports[`glimmer-utils maybeBlockParamDefinition able to handle single fiew params 1`] = `
Object {
  "getParentPath": [Function],
  "index": 1,
  "name": "boo",
  "type": "BlockParam",
}
`;

exports[`glimmer-utils maybeBlockParamDefinition able to handle single param 1`] = `
Object {
  "getParentPath": [Function],
  "index": 0,
  "name": "items",
  "type": "BlockParam",
}
`;
